packages/redis/src/RedisStreamSource.ts

```diff
@@ -0,0 +1,142 @@
+import {
+    IInputSource,
+    IRequireInitialization,
+    IDisposable,
+    MessageRef,
+    IComponentContext,
+    DefaultComponentContext,
+    Lifecycle,
+    makeLifecycle,
+} from "@walmartlabs/cookie-cutter-core";
+import { Span, Tags, Tracer } from "opentracing";
+import { generate } from "shortid";
+
+import { IRedisInputStreamOptions, IRedisClient, redisClient } from ".";
+import { RedisOpenTracingTagKeys } from "./RedisClient";
+
+export class RedisStreamSource implements IInputSource, IRequireInitialization, IDisposable {
+    private done: boolean = false;
+    private client: Lifecycle<IRedisClient>;
+    private tracer: Tracer;
+    private spanOperationName: string = "Redis Input Source Client Call";
+    private consumerId: string;
+
+    constructor(private readonly config: IRedisInputStreamOptions) {
+        this.tracer = DefaultComponentContext.tracer;
+        this.consumerId = generate();
+    }
+
+    public async *start(): AsyncIterableIterator<MessageRef> {
+        while (!this.done) {
+            const span = this.tracer.startSpan(this.spanOperationName);
+
+            this.spanLogAndSetTags(
+                span,
+                this.config.db,
+                this.config.readStream,
+                this.config.consumerGroup
+            );
+
+            const pendingMessages = await this.client.xPending(
+                span.context(),
+                this.config.readStream,
+                this.config.consumerGroup,
+                5
+            );
+
+            const expiredIdleMessages = pendingMessages.filter(
+                ([, , idleTime]) => idleTime > this.config.idleTimeoutMs
+            );
```

Comment:
Since we're only querying for a maximum of 5 pending items, it's possible that our post-query filter here yields no messages, but there could still exist unacknowledged messages with idle times greater than `this.config.idleTimeoutMs`. In order to be sure that we have checked _all_ of the pending messages, we would need to continue to call `XPENDING` until it returned less than `COUNT` messages, adjusting the value of `start` for each subsequent call (instead of always setting it to `-`).

More generally, I'm curious as to why we're using `XPENDING` instead of invoking `XREADGROUP` with an `ID` of 0 to fetch the unacknowledged messages.

From [the docs](https://redis.io/commands/xreadgroup#differences-between-xread-and-xreadgroup):
> The ID to specify in the **STREAMS** option when using `XREADGROUP` can be one of the following two:
>
>  * The special > ID, which means that the consumer wants to receive only messages that were _never delivered to any other consumer_. It just means, give me new messages.
>  *  Any other ID, that is, 0 or any other valid ID or incomplete ID (just the millisecond time part), will have the effect of returning entries that are pending for the consumer sending the command with IDs greater than the one provided. So basically if the ID is not >, then the command will just let the client access its pending entries: messages delivered to it, but not yet acknowledged.

The only reason I could see to query for stale messages across all consumers by using `XPENDING` is to recover from a multi-instance consumer group permanently losing a member (e.g. an operational decision is made to scale down the group to fewer members). 🤔

Reply:
I am also going to change that hardcoded number of 5 to be a config option `idleMessageBatchSize`, but you are correct that this functionality is to recover if a consumer dies.  While this functionality is def. geared towards multi consumer recovery, this will also successfully recover dropped messages from a single consumer.

Reply:
My apologies, I should have retracted the second half of this comment after I learned that we were auto-generating consumer names.

I still think we'll need to do more than make `idleMessageBatchSize` configurable. I think it needs to be called repeatedly, with increasing lower bound values (as of now we always pass `-`), until `XPENDING` returns fewer than `idleMessageBatchSize` messages.